rs/sns/governance/src/governance.rs

```diff
@@ -4706,6 +4841,96 @@ impl Governance {
             }
         }
     }
+    /// Checks if an automatic upgrade is needed and initiates it.
+    /// An automatic upgrade is needed if `target_version` is set to a future version on the upgrade path
+    async fn initiate_upgrade_if_behind_target_version(&mut self) {
+        if self.check_no_upgrades_in_progress().is_err() {
+            // An upgrade is already in progress
+            return;
+        }
+
+        let Some(deployed_version) = self.proto.deployed_version.clone() else {
+            return;
+        };
+
+        let Some(target_version) = self.proto.target_version.clone() else {
+            return;
+        };
+
+        let Some(CachedUpgradeSteps {
+            upgrade_steps: Some(Versions {
+                versions: upgrade_steps,
+            }),
+            ..
+        }) = &self.proto.cached_upgrade_steps
+        else {
+            log!(ERROR, "Cached upgrade steps set to None");
+            return;
+        };
+
+        // Find the current position of the deployed version
+        let Some(deployed_position) = upgrade_steps.iter().position(|v| v == &deployed_version)
+        else {
+            log!(ERROR, "Deployed version is not on the upgrade path");
+            self.invalidate_cached_upgrade_steps();
+            return;
+        };
+
+        // Find the target position of the target version
+        let Some(target_position) = upgrade_steps.iter().position(|v| v == &target_version) else {
```

Comment:
I've seen this twice, and it feels like it should live in a function.

Reply:
Did you mean the whole thing including the `else {...}`, or just the `upgrade_steps.iter().position(|v| v == &target_version)` part?

Reply:
Sort of feels like it could be a method on CachedUpgradeSteps to get the position of a given version in the upcoming steps...